teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,372 @@
+
+### Related Work
+
+Similar expression evaluations are supported in other CI/CD tools in the industry:
+- [Argo][Argo]: `Workflows` support conditional execution using a when property to specify whether to run or skip a `Step` ([example][argo example]).
+- [GitHub Actions][GitHub Actions]: supports evaluating expressions in `if` of a `step` to enable conditional execution.
+- [Jenkins][Jenkins]: executes the stage when the specified Groovy expression evaluates to true.
+- [Spinnaker][Spinnaker]: Uses string expressions, `Stages` only run when expressions evaluate to `True`.
+
+
+## Proposal
+
+This doc proposes to add a `string` type field `cel` under `WhenExpression`. Like [CelCustomRun][CELCustomRun], we propose to use [cel-go][cel-go] for the expression evaluation. This will allow users to craft any valid CEL expression as defined by the [cel-spec][cel-spec] language definition.
+
+
+### Syntax
+
+In this section we will show how CEL can support the [use cases](#use-cases).
+
+#### Use CEL to replace current WhenExpression
+```yaml
+# current WhenWxpressions
+when:
+  - input: "foo"
+    operator: "in"
+    values: ["foo", “bar”]
+  - input: "duh"
+    operator: "notin"
+    values: ["foo", “bar”]
+
+# with cel
+when:
+  - cel: "'foo' in ['foo', 'bar']”
+  - cel: "'duh' not in ['foo', 'bar']”
+```
+
+**Note:** The `WhenExpressions` contains a list of `WhenExpression`, if all `WhenExpressions` are true it will return true and allow execution. So we’re using AND to evaluate the list of `WhenExpressions`. And this will remain the same for this proposal. Each existing `WhenExpression` can be replaced with a CEL expression.
+
+#### Use CEL to Support ANY operator
+
+This is the workaround from [#3591][#3591],
+```yaml
+  when:
+    - input: "$(params.param1)$(params.param2)"
+      operator: notin
+      values: [""]
+```
+
+With CEL we can use a more intuitive expression to support it.
+```yaml
+  when:
+    - cel: "'$(params.param1)' != '' || '$(params.param2)' != ''"
+```
+
+#### Use CEL to Support Numeric Comparisons
+
+This is not possible with current `WhenExpression`. With CEL we can do:
+
+```yaml
+  when:
+    - cel: "'$(tasks.unit-test.results.test-coverage)' >= 0.9"
+```
+
+[Note:][numeric note] currently there are no automatic arithmetic conversions for the numeric types (int, uint, and double), so the emitted results must be the same numeric type to make sure the evaluation works.
+
+#### Use CEL to Support Pattern Matching:
+
+This is not possible with current `WhenExpression`. With CEL we can do:
+
+```yaml
+  when:
+    - cel: '$(params.branch)'.matches('release/.*')"
+```
+
+### Variables Substitution
+
+The CEL in When Expressions should support current Tekton’s Params and Results [string substitutions](https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#variables-available-in-a-pipeline). Array and object substitutions are not supported because Tekton doesn’t support whole array/object replacements in a string.
```

Comment:
is this true? `values` is an array and it does honor the star notation, isn't it? 

Its reasonable to implement only string params with CEL but are we supporting indexing into an array or referencing an object param using a key in CEL expressions?

Reply:
Yes, I should clarify that we cannot reference to a whole array or object.

Reply:
I remember there is an issue that we cannot reference param in "script", the case where a long string contains param reference. 

I will remove this statement, and rephrase that we will support string replacements first

Reply:
Removed this statement